motif/match.go

```diff
@@ -0,0 +1,304 @@
+				}
+			}
+			if !needNewKey {
+				if currScore, inKmerHash = kmerHash[currKey]; inKmerHash { //if we get a hit from the current key in the kmer hash.
+					if outputAsProportion {
+						currScore = currScore / consensusScore
+					}
+					currBed = bed.Bed{Chrom: records[currChrom].Name,
+						ChromStart:        currPos - motifLen,
+						ChromEnd:          currPos,
+						Name:              motifName,
+						Score:             0,
+						Strand:            strand,
+						FieldsInitialized: 7,
+						Annotation:        []string{fmt.Sprintf("%f", currScore)}}
+					bed.WriteBed(out, currBed)
+				}
+			}
+		}
+	}
+}
+
+// getNewKey generates a 2bit encoded kmer sequence as a uint64 from a multiple alignment starting at an input alnPos.
+// for a pairwise alignment, index should be 0 for target and 1 for query.
+// returns the uint64 seq, an int corresponding to the ending alnPos, and a bool that returns true if a key was generated without
+// reaching the end of the alignment.
+func getNewKey(record fasta.Fasta, alnPos int, motifLen int) (uint64, int, bool) {
+	var answer uint64
+	var motifPos = 0
+	for motifPos < motifLen {
+		if alnPos >= len(record.Seq) {
+			return 0, 0, false
+		}
+		switch record.Seq[alnPos] {
+		case dna.N: //if we see an N, we skip this base and clear out answer and the partial key.
+			alnPos++
+			motifPos = 0
+			answer = 0
+		case dna.Gap: //keep searching, motifs can span gaps, which are ignored.
+			alnPos++
+		case dna.A:
+			answer = answer << 2 //left shift two bases, clearing a space for the new base.
+			answer = answer | 0  //append 00 in last two spots with bitwise OR.
+			motifPos++
+		case dna.C:
+			answer = answer << 2
+			answer = answer | 1
+			motifPos++
+		case dna.G:
+			answer = answer << 2
+			answer = answer | 2
+			motifPos++
+		case dna.T:
+			answer = answer << 2
+			answer = answer | 3
+			motifPos++
+		default:
+			log.Fatalf("Unrecognized base.")
+		}
+	}
+	return answer, alnPos, true
+}
+
+// rankTensorElement encodes the value and base for a rankTensor at a particular row and column.
+// In three dimensions, it is useful to consider the Value and Base as two layers of the rankTensor.
+type rankTensorElement struct {
+	Value float64
+	Base  dna.Base
+}
+
+// initializeRankTensor turns a PositionMatrix into a rank-ordered position weight tensor.
+func initializeRankTensor(p PositionMatrix) [][]rankTensorElement {
+	var row, column int
+	var currMaxRow, currRank int
+	var currMaxValue float64
+	var answer = make([][]rankTensorElement, 4)
+	for row = 0; row < 4; row++ {
+		answer[row] = make([]rankTensorElement, len(p.Mat[row]))
+		for column = 0; column < len(p.Mat[row]); column++ {
+			answer[row][column] = rankTensorElement{Value: p.Mat[row][column], Base: dna.Base(row)} //cast row to dna.Base (0 -> A, 1 -> C, 2 -> G, 3 -> T)
+		}
+	}
+	//sort matrix columns by rank
+	for column = 0; column < len(p.Mat[0]); column++ {
+		for currRank = 0; currRank < 3; currRank++ {
```

Comment:
Style: This is a small consideration but what is the most efficient way to sort the matrix? I was thinking maybe only 1 for loop would be necessary?

Reply:
This is performing Insertion Sort, which runs in O(n^2). Indeed not the fastest, many sorting algorithms run in O(nlogn). However, since N is guaranteed to be 4, the difference should be minimal. I chose this sorting method for ease of programming and readability here